File: bin/Data/AUI/Periodic/LuaScripts/code_awindow.lua
-- AWindow application source code


function setup_uiwindow( layout, uiview )
    local dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do
        SubscribeToEvent(awid, "WidgetEvent", "HandleUiwindowEvent")
    end
end

function HandleUiwindowEvent(eventType, eventData)
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
        return
    end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uiwindowcode" then
            AppLog( "AWindow support : " .. widget:GetId() .. " was pressed " )
            ViewCode ( "LuaScripts/code_awindow.lua", widget:GetParent())
        end
        if widget:GetId() ==  "uiwindowlayout" then
            AppLog( "AWindow support : " .. widget:GetId() .. " was pressed ")
            ViewCode ( "Scenes/layout_awindow.ui.txt", widget:GetParent())
        end

        if widget:GetId() == "windowdemo" then
            AppLog( "AWindow action : " .. widget:GetId() .. " was pressed " )
            local win1 = AWindow:new()
            win1:SetSettings ( UI_WINDOW_SETTINGS_DEFAULT )
            win1:SetText("AWindow demo (a login dialog)")
            win1:Load("Scenes/login_dialog.ui.txt")
            win1:ResizeToFitContent()
            uiview:AddChild(win1)
            win1:Center()
            local login = win1:GetWidget("login")
            SubscribeToEvent(login, "WidgetEvent", "HandleUiwindowEvent" )
            local cancel = win1:GetWidget("cancel")
            SubscribeToEvent(cancel, "WidgetEvent", "HandleUiwindowEvent" )
        end
        if widget:GetId() == "login" then
            AppLog( "AWindow action : " .. widget:GetId() .. " was pressed " )
            local mywin = FindTheWindowParent(widget)
            if mywin ~= nil then
                local w1 = tolua.cast( mywin, "AWindow")
                if w1 ~= nil then
                    mywin:Close()
                end
            end
        end
        if widget:GetId() == "cancel" then
            AppLog( "AWindow action : " .. widget:GetId() .. " was pressed " )
            local mywindow = FindTheWindowParent(widget)
            if mywindow ~= nil then
                local w1 = tolua.cast( mywindow, "AWindow")
                if w1 ~= nil then
                    w1:Close()
                end
            end
        end
        if widget:GetId() == "windowdemo1" then
            AppLog( "AWindow action : " .. widget:GetId() .. " was pressed " )
            local win2 = AWindow:new()
            win2:SetSettings ( UI_WINDOW_SETTINGS_DEFAULT )
            win2:SetText("AWindow demo (a table)" )
            win2:Load("Scenes/sheet.ui.txt")
            win2:ResizeToFitContent()
            uiview:AddChild(win2)
            win2:Center()
        end
        if widget:GetId() == "windowdemo2" then
            AppLog( "AWindow action : " .. widget:GetId() .. " was pressed " )
            local win3 = AWindow:new()
            win3:SetAxis(UI_AXIS_X)
            win3:SetSettings ( UI_WINDOW_SETTINGS_DEFAULT )
            win3:SetText(" ")
            win3:Load("Scenes/sheet.ui.txt")
            win3:ResizeToFitContent()
            uiview:AddChild(win3)
            win3:Center()
        end
        if widget:GetId() == "windowdemo3" then
            AppLog( "AWindow action : " .. widget:GetId() .. " was pressed " )
            local win4 = AWindow:new()
            win4:SetSettings ( UI_WINDOW_SETTINGS_DEFAULT )
            win4:SetId ("WinCloseOverride" )
            win4:SetText("AWindow with Close Override action" )
            win4:Load("Scenes/sheet.ui.txt")
            win4:SetCloseDelegate(true)
            SubscribeToEvent(win4, "AWidgetEditCanceled", "HandleCloseOverrideEvent")
            win4:ResizeToFitContent()
            uiview:AddChild(win4)
            win4:Center()
        end
    end
end

function HandleCloseOverrideEvent(eventType, eventData)
    local widget = eventData["Widget"]:GetPtr("AWidget")
    if widget == nil then
        return
    end
    local mess4 = AMessageWindow:new( uiview, "mymess4")
    SubscribeToEvent(mess4, "WidgetEvent", "HandleCloseYesNoEvent" )
    mess4:Show( "Close Window Request", "Do you really want to close this window?", UI_MESSAGEWINDOW_SETTINGS_YES_NO, false, 0, 0)
end

function HandleCloseYesNoEvent(eventType, eventData)
    local refid = eventData["RefID"]:GetString()
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
        return
    end

    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if refid == "TBMessageWindow.yes" then
            local mywin = widget:FindWidget("WinCloseOverride")
            if mywin ~= nil then
                local wcx = tolua.cast(mywin, "AWindow" )
                if wcx ~= nil then
                    wcx:Close()
                end
            end
        end
        if refid == "TBMessageWindow.no" then
        end
    end
end

File: Source/Samples/95_Periodic/code_awindow.cpp
// AWindow application source code
#include <Urho3D/AUI/AMessageWindow.h>
#include <Urho3D/AUI/AWindow.h>
#include <Urho3D/AUI/AView.h>
#include "PeriodicApp.h"

void PeriodicApp::setup_awindow( AWidget *layout, AView *uiview )
{
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAwindowEvent ));
}

void PeriodicApp::HandleAwindowEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uiwindowcode" )
        {
            AppLog( "AWindow support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_awindow.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uiwindowlayout" )
        {
            AppLog( "AWindow support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_awindow.ui.txt", widget->GetParent() );
        }

        if (widget->GetId() == "windowdemo" )
        {
            AppLog( "AWindow action : " + widget->GetId() + " was pressed " );
            AView *someview = widget->GetView();
            AWindow *window = new AWindow(context_);
            window->SetSettings ( UI_WINDOW_SETTINGS_DEFAULT );
            window->SetText("AWindow demo (a login dialog)");
            window->Load("Scenes/login_dialog.ui.txt");
            window->ResizeToFitContent();
            someview->AddChild(window);
            window->Center();
            AWidget *login = window->GetWidget("login");
            SubscribeToEvent(login, E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAwindowEvent ));
            AWidget *cancel = window->GetWidget("cancel");
            SubscribeToEvent(cancel, E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAwindowEvent ));
        }
        if (widget->GetId() == "login" )
        {
            AppLog( "AWindow action : " + widget->GetId() + " was pressed " );
            AWindow *mywindow = static_cast<AWindow *>(FindTheWindowParent(widget));
            if (mywindow)
                mywindow->Close();
        }
        if (widget->GetId() == "cancel" )
        {
            AppLog( "AWindow action : " + widget->GetId() + " was pressed " );
            AWindow *mywindow = static_cast<AWindow *>(FindTheWindowParent(widget));
            if (mywindow)
                mywindow->Close();
        }
        if (widget->GetId() == "windowdemo1" )
        {
            AppLog( "AWindow action : " + widget->GetId() + " was pressed " );
            AView *someview = widget->GetView();
            AWindow *window = new AWindow(context_);
            window->SetSettings ( UI_WINDOW_SETTINGS_DEFAULT );
            window->SetText("AWindow demo (a table)" );
            window->Load("Scenes/sheet.ui.txt");
            window->ResizeToFitContent();
            someview->AddChild(window);
            window->Center();
        }
        if (widget->GetId() == "windowdemo2" )
        {
            AppLog( "AWindow action : " + widget->GetId() + " was pressed " );
            AView *someview = widget->GetView();
            AWindow *window = new AWindow(context_);
            window->SetAxis(UI_AXIS_X);
            window->SetSettings ( UI_WINDOW_SETTINGS_DEFAULT );
            window->SetText(" ");
            window->Load("Scenes/sheet.ui.txt");
            window->ResizeToFitContent();
            someview->AddChild(window);
            window->Center();
        }
        if (widget->GetId() == "windowdemo3" )
        {
            AppLog( "AWindow action : " + widget->GetId() + " was pressed " );
            AView *someview = widget->GetView();
            AWindow *window = new AWindow(context_);
            window->SetId ("WinCloseOverride" );
            window->SetSettings ( UI_WINDOW_SETTINGS_DEFAULT );
            window->SetText("AWindow with Close Override action" );
            window->Load("Scenes/sheet.ui.txt");
            window->ResizeToFitContent();
            window->SetCloseDelegate(true);
            SubscribeToEvent(window, E_AWIDGETEDITCANCELED, URHO3D_HANDLER(PeriodicApp,HandleCloseOverrideEvent));
            someview->AddChild(window);
            window->Center();
        }
    }
}

void PeriodicApp::HandleCloseOverrideEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace AWidgetEditCanceled;
    AWidget* widget = static_cast<AWidget*>(eventData[P_WIDGET].GetPtr());
    if ( widget == NULL ) return;

    AView *someview = widget->GetView();
    AMessageWindow *mess4 = new AMessageWindow(context_, someview, "mymess4");
    SubscribeToEvent(mess4, E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleCloseYesNoEvent ));
    mess4->Show( "Close Window Request", "Do you really want to close this window?", UI_MESSAGEWINDOW_SETTINGS_YES_NO, 0, 0, 0);
}

void PeriodicApp::HandleCloseYesNoEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    AWindow* mywin = static_cast<AWindow*>(widget->FindWidget("WinCloseOverride"));
    String refid = eventData[P_REFID].GetString();
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (refid == "TBMessageWindow.yes" )
        {
            AppLog( "AWindow event : " + refid + " closed the AMessageWindow");
            if (mywin)
                mywin->Close();
        }
        if (refid == "TBMessageWindow.no" )
        {
            AppLog( "AWindow event : " + refid + " closed the AMessageWindow");
            //wait for it ...
        }
    }
}

